From ca4eea119f2c5fed85665d7e2386cde843fbb554 Mon Sep 17 00:00:00 2001
From: Philippe Canal <pcanal@fnal.gov>
Date: Thu, 17 Oct 2013 05:07:29 -0500
Subject: [PATCH] Fix thread safety of TThread::TThread(Long_t id)

Changing fId is a common ressource (see TThread::GetThread)
and thus its update must be locked.

(cherry picked from commit c85bd1176dee1629ac1beb0d7f57003198dd281a)
---
 core/thread/src/TThread.cxx |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/core/thread/src/TThread.cxx b/core/thread/src/TThread.cxx
index 0fc81bb..bfb1851 100644
--- a/core/thread/src/TThread.cxx
+++ b/core/thread/src/TThread.cxx
@@ -246,9 +246,14 @@ TThread::TThread(Long_t id)
    fPriority  = kNormalPriority;
    fThreadArg = 0;
    Constructor();
+
+   // Changing the id must be protected as it will be look at by multiple
+   // threads (see TThread::GetThread)
+   ThreadInternalLock();
    fNamed     = kFALSE;
    fId = (id ? id : SelfId());
    fState = kRunningState;
+   ThreadInternalUnLock();
 
    if (gDebug)
       Info("TThread::TThread", "TThread attached to running thread");
-- 
1.7.4.1

# bla bla
